### PR TITLE
storageccl: add setting for GCS key

### PR DESCRIPTION
**Pull request by madelynnblue**
Previously the `gs` storage scheme for backup/restore could only use
instance (implicit) auth. The other schemes allowed to pass auth
secrets in the URLS. However Google's auth stuff is much longer
than just an id and secret key: it is a JSON file with various
fields. Marshaling that to base64 to put into a URL would have been a
bit overblown. Also, being forced to hardcode those auth keys forced
operators to execute queries with them even when they would prefer
to have the keys more hidden.

After the recent change to disallow non-root from even viewing cluster
settings, we can now put these auth keys into the cluster settings
and configure the storage providers to use them. We are starting with
Google because it was the only one lacking before.

Add a cluster setting `cloudstorage.gs.default.key` that holds the
contents of a Google Cloud Storage JSON key file. The namespacing is:
`cloudstorage.<scheme>.<name>.<key/id/secret/etc>`. The `name` field
here will start as `default` for all schemes. We may allow users to
add their own custom-named keys, but that will require some changes
to the settings infra, and so may be done later. But the namespacing
is done as is to allow this future enhancement.

Add an AUTH parameter to GCS export storages. There are three valid
values. If the AUTH parameter is not present, the storage will check
if the key setting exists and use it if it does, otherwise fall back
to auth from the instance/environment (hereafter called implicit). A
value of `default` only checks the key, and a value of `implicit`
only uses implicit auth.

See #18822

### CODE REVIEW

**Comment by dt** [comment]
@bdarnell brought up some concerns about saving credentials in settings on the feature issue #18822 that we should probably resolve before moving forward here.

**Comment by dt** [comment]
#19031

File: pkg/ccl/storageccl/export_storage.go [hunk 1]
```diff
@@ -45,6 +50,19 @@ const (
...
+	// key in a URI.
+	AuthParam         = "AUTH"
+	authParamImplicit = "implicit"
+	authParamDefault  = "default"
+
+	cloudstoragePrefix           = "cloudstorage"
+	cloudstorageGS               = cloudstoragePrefix + ".gs"
+	cloudstorageDefault          = ".default"
+	cloudstorageGSDefault        = cloudstorageGS + cloudstorageDefault
+	cloudstorageGSJsonkey        = ".jsonkey"
```
Comment (dt) [inline pkg/ccl/storageccl/export_storage.go]: 🚲🏡: arguably we don't need `jsonkey`, but it keep symmetry with the other providers where we'll have suffixes for `.key` and `.secret` or whatever.

File: pkg/ccl/storageccl/export_storage.go [hunk 2]
```diff
@@ -427,11 +454,41 @@ func (g *gcsStorage) Conf() roachpb.ExportStorage {
...
+		}
+		if key != "" {
+			source, err := google.JWTConfigFromJSON([]byte(key), scope)
+			if err != nil {
+				return nil, errors.Wrap(err, "creating GCS oauth token source")
+			}
+			opts = append(opts, option.WithTokenSource(source.TokenSource(ctx)))
+		}
+	case authParamImplicit:
+		// do nothing; use implicit params
```
Comment (dt) [inline pkg/ccl/storageccl/export_storage.go]: nit: i'd link to SDK doc on behavior.

File: pkg/ccl/cliccl/load.go [hunk 3]
```diff
@@ -169,7 +169,7 @@ func runLoadShow(cmd *cobra.Command, args []string) error {
 			return err
 		}
 	}
-	desc, err := sqlccl.ReadBackupDescriptorFromURI(ctx, basepath)
+	desc, err := sqlccl.ReadBackupDescriptorFromURI(ctx, basepath, nil /* cluster.Settings */)
```
Comment (dt) [inline pkg/ccl/cliccl/load.go]: here+elsewhere: if you felt so inclined, I'd love to see a alias like `settings.NoSetttingsValues = nil` so we could just use that in place of the commented nil args`nil /* ...*/`

**Review by dt — APPROVED** [review]
We also have `TestCloudBackupRestoreGoogleCloudStorage` in sqlccl's tests if you wanted to exercise from SQL too.

File: pkg/ccl/cliccl/load.go (hunk 3 above)
Comment (mjibson) [reply pkg/ccl/cliccl/load.go]: Done.

File: pkg/ccl/storageccl/export_storage.go (hunk 1 above)
Comment (mjibson) [reply pkg/ccl/storageccl/export_storage.go]: Changed to `.key`

File: pkg/ccl/storageccl/export_storage.go (hunk 2 above)
Comment (mjibson) [reply pkg/ccl/storageccl/export_storage.go]: Done.

**Comment by mjibson** [comment]
RFAL. Since the cluster settings stuff is in, I think this can go in now.

**Comment by mjibson** [comment]
ping @dt

**Comment by dt** [comment]
:lgtm:

---

Reviewed 2 of 11 files at r1, 7 of 11 files at r3, 1 of 4 files at r4.
Review status: 5 of 16 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19018#-:-KwBpWToeOkE7T7Zt0hR:bnfp4nl)*

**Pull request merged**